File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 0;
        int j = height.size() - 1;
        int leftMax = height[0];
        int rightMax = height[j];
        int trappedWater = 0;
        
        while (i < j) {
            if (leftMax <= rightMax) {
                trappedWater += leftMax - height[i];
                i++;
                leftMax = std::max(leftMax, height[i]);
            } else {
                trappedWater += rightMax - height[j];
                j--;
                rightMax = std::max(rightMax, height[j]);
            }
        }
        return trappedWater;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut i = 0;
        let mut j = height.len() - 1;
        let mut left_max = height[0];
        let mut right_max = height[j];
        let mut trapped_water = 0;

        while i < j {
            if left_max <= right_max {
                trapped_water += (left_max - height[i]);
                i += 1;
                left_max = std::cmp::max(left_max, height[i]);
            } else { 
                trapped_water += (right_max - height[j]);
                j -= 1;
                right_max = std::cmp::max(right_max, height[j]);
            }
        }
        trapped_water
    }
}